pkg/topology/kademlia/kademlia.go

```diff
@@ -1204,13 +1269,41 @@ func (k *Kad) EachNeighborRev(f topology.EachPeerFunc) error {
 }
 
 // EachPeer iterates from closest bin to farthest.
-func (k *Kad) EachPeer(f topology.EachPeerFunc) error {
-	return k.connectedPeers.EachBin(f)
+func (k *Kad) EachPeer(f topology.EachPeerFunc, filter topology.Filter) error {
+	return k.connectedPeers.EachBin(func(addr swarm.Address, po uint8) (bool, bool, error) {
+		if filter.Reachable && k.peerFilter(addr) {
+			return false, false, nil
+		}
+		return f(addr, po)
+	})
 }
 
 // EachPeerRev iterates from farthest bin to closest.
-func (k *Kad) EachPeerRev(f topology.EachPeerFunc) error {
-	return k.connectedPeers.EachBinRev(f)
+func (k *Kad) EachPeerRev(f topology.EachPeerFunc, filter topology.Filter) error {
+	return k.connectedPeers.EachBinRev(func(addr swarm.Address, po uint8) (bool, bool, error) {
+		if filter.Reachable && k.peerFilter(addr) {
+			return false, false, nil
+		}
+		return f(addr, po)
+	})
+}
+
+// SetPeerReachability sets the peer reachability status.
+func (k *Kad) Reachable(addr swarm.Address, status p2p.ReachabilityStatus) {
+	k.collector.Record(addr, im.PeerReachability(status))
```

Comment:
I'd consider adding a `Tracef` here telling that the node reachability status has been reported, etc